bootstrap/main.py

```diff
@@ -0,0 +1,194 @@
+# Copyright 2018 Open Source Foundries Limited.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+'''West's bootstrap/wrapper script.
+'''
+
+import argparse
+import os
+from os.path import join
+import importlib
+import platform
+import subprocess
+import sys
+
+MANIFEST = 'manifest'
+MANIFEST_DEFAULT = 'https://github.com/zephyrproject-rtos/manifest'
+MANIFEST_REV_DEFAULT = 'master'
+WEST = 'west'
+WEST_DEFAULT = 'https://github.com/zephyrproject-rtos/west'
+WEST_REV_DEFAULT = 'master'
+if platform.system() == 'Windows':
+    # On Windows, the sentinel is hidden with attributes, rather than a
+    # leading '.', and thus doesn't contain one.
+    WEST_HIDDEN_SENTINEL = 'west-top'
+else:
+    # Assume a POSIX OS with leading '.' for hidden files.
+    WEST_HIDDEN_SENTINEL = '.west-top'
+
+#
+# Helpers shared between init and wrapper mode
+#
+
+
+class WestError(RuntimeError):
+    pass
+
+
+class WestNotFound(WestError):
+    '''Neither the current directory nor any parent has a West installation.'''
+
+
+def find_west_topdir(start):
+    '''Find the top-level installation directory, starting at ``start``.
+
+    If none is found, raises WestNotFound.'''
+    if not os.path.isdir(start):
+        raise WestNotFound()
+
+    if WEST_HIDDEN_SENTINEL in os.listdir(start):
+        return start
+
+    dirname = os.path.dirname(start)
+    if start == dirname:
+        # / on POSIX, top level drive name on Windows.
+        raise WestNotFound()
+    else:
+        return find_west_topdir(dirname)
+
+
+def make_topdir_sentinel(directory):
+    '''Make the hidden sentinel that marks a West installation.'''
+    # The sentinel is a directory in case we ever need to stash
+    # anything inside. It could just as well be an empty file for now,
+    # but this is more future-proof.
+    sentinel = join(directory, WEST_HIDDEN_SENTINEL)
+    os.mkdir(sentinel)
+    # On Windows, we have to manually mark a file as hidden.
+    if platform.system() == 'Windows':
+        subprocess.check_call(['attrib', '+H', sentinel])
+
+
+def clone(url, rev, dest):
+    def repository_type(url):
+        if (url.startswith('http:') or url.startswith('https:') or
+                url.startswith('git:') or url.startswith('git+ssh:') or
+                url.startswith('file:')):
+            return 'GIT'
+        else:
+            return 'UNKNOWN'
+
+    if os.path.exists(dest):
+        msg = 'refusing to clone into existing location {}'.format(dest)
+        raise WestError(msg)
+
+    rtype = repository_type(url)
+    if rtype == 'GIT':
```

Comment:
`rtype` doesn't seem to be used elsewhere, so could do `if repository_type(url) == 'GIT':'.

Reply:
I was just planning ahead here for adding schemas for p4, svn, etc. as discussed at the kickoff. This way allows adding those as additional elif blocks, minimizing diff. But it'd be a trivial patch either way, so I'm fine with that.